OSDOCS 5806: Adding docs for using the secrets store CSI driver (#62385)
openshift/openshift-docs: modules/secrets-store-aws-secrets-manager.adoc

      hostNetwork: false
      containers:
        - name: provider-aws-installer
          image: public.ecr.aws/aws-secrets-manager/secrets-store-csi-driver-provider-aws:1.0.r2-50-g5b4aca1-2023.06.09.21.19
          imagePullPolicy: Always
          args:
              - --provider-volume=/etc/kubernetes/secrets-store-csi-providers
          resources:
            requests:
              cpu: 50m
              memory: 100Mi
            limits:
              cpu: 50m
              memory: 100Mi
          securityContext:
            privileged: true
          volumeMounts:
            - mountPath: "/etc/kubernetes/secrets-store-csi-providers"
              name: providervol
            - name: mountpoint-dir
              mountPath: /var/lib/kubelet/pods
              mountPropagation: HostToContainer
      tolerations:
      - operator: Exists
      volumes:
        - name: providervol
          hostPath:
            path: "/etc/kubernetes/secrets-store-csi-providers"
        - name: mountpoint-dir
          hostPath:
            path: /var/lib/kubelet/pods
            type: DirectoryOrCreate
      nodeSelector:
        kubernetes.io/os: linux

> ----

Reviewer: See the test case OCP-66032 , you lack a step `oc adm policy add-scc-to-user privileged -z csi-secrets-store-provider-aws -n openshift-cluster-csi-drivers`. If you don't want to use this command, you could append the YAML content generated by this command to these yaml snippets:
```
$ oc adm policy add-scc-to-user privileged -z csi-secrets-store-provider-aws -n openshift-cluster-csi-drivers --dry-run=client -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: system:openshift:scc:privileged
  namespace: openshift-cluster-csi-drivers
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system:openshift:scc:privileged
subjects:
- kind: ServiceAccount
  name: csi-secrets-store-provider-aws
  namespace: openshift-cluster-csi-drivers
```

+
[NOTE]
====
This configuration is modified from the configuration provided in the link:https://github.com/aws/secrets-store-csi-driver-provider-aws#installing-the-aws-provider[AWS documentation] so that it works properly with {product-title}.
====

.. Create the provider resources by running the following command:
+
[source,terminal]
----
$ oc apply -f aws-provider.yaml
----

. Create a secret provider class to define your secrets store provider:

.. Create a YAML file that defines the `SecretProviderClass` object:
+
.Example `secret-provider-class-aws.yaml`
[source,yaml]
----
apiVersion: secrets-store.csi.x-k8s.io/v1
kind: SecretProviderClass
metadata:
  name: my-aws-provider                   <1>
  namespace: my-namespace                 <2>
spec:
  provider: aws                           <3>
  parameters:                             <4>
    objects: |
      array:
        - |

>           objectName: secret1

Reviewer: See my test case(s), I did not use `array:` and `- |`,  https://github.com/aws/secrets-store-csi-driver-provider-aws or https://docs.aws.amazon.com/secretsmanager/latest/userguide/integrating_csi_driver.html  do not use either.
```
    objects: |
      - objectName: "xxiaSecret"
        objectType: "secretsmanager"
```
I am not sure if using `array:` and `- |` will work for AWS provider. But I can try it to confirm next day.

Reviewer: > I am not sure if using array: and - | will work for AWS provider. But I can try it to confirm next day.

Today I tried, while below format works well for AWS provider:
```
    objects: |
      - objectName: xxiaSecret
        objectType: secretsmanager
```
The format as your current PR does not work well for AWS provider:
```
spec:
  parameters:
    objects: |
      array:
        - |
          objectName: xxiaSecret
          objectType: secretsmanager
  provider: aws
```
With the format as your current PR, the pods is stuck in `ContainerCreating` with error `Failed to load SecretProviderClass: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal object into Go value of type []*provider.SecretDescriptor`:
```
$ oc get po
NAME                               READY   STATUS              RESTARTS   AGE
hello-openshift-84c76c5b89-9h7sc   0/1     ContainerCreating   0          11s

$ oc describe po hello-openshift-84c76c5b89-9h7sc
...
Events:
  Type     Reason       Age               From               Message
  ----     ------       ----              ----               -------
  Normal   Scheduled    17s               default-scheduler  Successfully assigned xxia-proj/hello-openshift-84c76c5b89-9h7sc to ip-10-0-81-84.ap-northeast-2.compute.internal
  Warning  FailedMount  1s (x6 over 17s)  kubelet            MountVolume.SetUp failed for volume "secrets-store-inline" : rpc error: code = Unknown desc = failed to mount secrets store objects for pod xxia-proj/hello-openshift-84c76c5b89-9h7sc, err: rpc error: code = Unknown desc = Failed to load SecretProviderClass: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal object into Go value of type []*provider.SecretDescriptor
```
So please change the format for AWS provider.

Author: Thanks, updated to:

```
    objects: |
      - objectName: "testSecret"
        objectType: "secretsmanager"
```

>           objectType: secret

Reviewer: See my test case(s), `objectType` should be `secretsmanager`. Also you can see https://github.com/aws/secrets-store-csi-driver-provider-aws or https://docs.aws.amazon.com/secretsmanager/latest/userguide/integrating_csi_driver.html , `objectType` uses `secretsmanager`.

Author: Thanks, I updated this to be secretsmanager throughout the AWS procedure (including the verification steps)

----
<1> Specify the name for the provider secret provider class.

> <2> Specify the namespace for the provider secret provider class.

Reviewer: ```suggestion
<2> Specify the namespace for the secret provider class.
```

<3> Specify the provider as `aws`.
<4> Specify the provider-specific configuration parameters.

.. Create the `SecretProviderClass` object by running the following command:
+
[source,terminal]
----
$ oc create -f secret-provider-class-aws.yaml
----

> . Create a deployment to use this secret provider class:

Reviewer: See the test case OCP-66032, you lack a step "`Grant the serviceaccount permission of reading the AWS secret object`". This is required for the deployment's pod to mount the secrets-store.csi.k8s.io CSI volume successfully to get Running status.

Author: Thanks, I added the steps for this. I had to generalize the input (for example, remove the "xxia" test values. Please let me know if I did everything correctly here

+
[source,terminal]
----
$ oc create -f secret-provider-class-aws.yaml
----

. Create a deployment to use this secret provider class:

.. Create a YAML file that defines the `Deployment` object:
+
.Example `deployment.yaml`
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-aws-deployment                              <1>
  namespace: my-namespace                              <2>
spec:
  replicas: 1
  selector:
    matchLabels:
      app: my-storage
  template:
    metadata:
      labels:
        app: my-storage
    spec:
      containers:
      - name: busybox
        image: k8s.gcr.io/e2e-test-images/busybox:1.29
        command:
          - "/bin/sleep"
          - "10000"
        volumeMounts:
        - name: secrets-store-inline
          mountPath: "/mnt/key-store"
          readOnly: true
      volumes:
        - name: secrets-store-inline
          csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: "my-aws-provider" <3>
----
<1> Specify the name for the deployment.
<2> Specify the namespace deployment.

> // TODO: Does this have to be the same ns as the secret provider class?

Author: @xingxingxia @ropatil010 I used the same namespace for the deployment as the secret provider class object. Do they have to be in the same namespace, or could this deployment be in some other namespace?

Reviewer 2: Yeah correct, it should be in the same namespace as per my understanding. 
cc: @xingxingxia

Reviewer: Yes, https://secrets-store-csi-driver.sigs.k8s.io/concepts#secretproviderclass : "`NOTE: The SecretProviderClass needs to be created in the same namespace as the pod.`"

Author: Great I'll work that into the callout, thanks!

<3> Specify the name of the secret provider class.

.. Create the `Deployment` object by running the following command:
+
[source,terminal]
----
$ oc create -f deployment.yaml
----

.Verification

* Verify that you can access the secrets from AWS Secrets Manager in the pod volume mount:

.. List the secrets in the pod mount:
+
[source,terminal]
----
$ oc exec busybox-<hash> -n my-namespace -- ls /mnt/key-store/
----
+
.Example output
[source,terminal]
----
tls-cert
----

.. View a secret in the pod mount:
+
[source,terminal]
----
$ oc exec busybox-<hash> -n my-namespace -- cat /mnt/key-store/tls-cert
----
+
.Example output
[source,terminal]
----
my-secret-value
----

> // TODO: Is this okay example output, or should it be something more specific to our example?

Author: @xingxingxia @ropatil010 Does this basic example output look okay for the secret example we've used, or should this be something more detailed?

Reviewer: Yes okay.

----
$ oc adm policy add-scc-to-user privileged -z csi-secrets-store-provider-aws -n openshift-cluster-csi-drivers
----

.. Create the provider resources by running the following command:
+
[source,terminal]
----
$ oc apply -f aws-provider.yaml
----

. Grant permission to allow the service account to read the AWS secret object:

.. Create a directory to contain the credentials request by running the following command:
+
[source,terminal]
----
$ mkdir credentialsrequest-dir-aws
----

.. Create a YAML file with the following configuration for the credentials request:
+
.Example `credentialsrequest.yaml` file
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: aws-provider-test
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AWSProviderSpec
    statementEntries:
    - action:
      - "secretsmanager:GetSecretValue"
      - "secretsmanager:DescribeSecret"
      effect: Allow
      resource: "arn:*:secretsmanager:*:*:secret:testSecret-??????"
  secretRef:
    name: aws-creds
    namespace: test-namespace
  serviceAccountNames:
  - aws-provider
----
+

> // TODO `resource` value: should we say something about this? I changed it from "secret:xxiaSecret" to "secret:testSecret". What are the "??????" supposed to be?

Author: @xingxingxia I'm not sure how to make the value for "resource" more generic. I changed "xxiaSecret" to "testSecret", which is what I use later on in the procedure. But I don't know what the question marks represent here, or if there's something else we need to tell users to replace.

Reviewer: I copied from https://github.com/aws/secrets-store-csi-driver-provider-aws which says `To simplify this example we use wild card matches`.
It is up to what the resource type's full ARN is generated by AWS. For secretsmanager secret, from the case OCP-66032's step `aws --region "$REGION" secretsmanager  create-secret --name xxiaSecret --secret-string ` and its output `"ARN": "arn:aws:secretsmanager:us-east-2:hiddenaccount:secret:xxiaSecret-AzC4eU",`, the full ARN of secretsmanager secret has a random 6-byte suffix, so it uses six ? symbols.

Author: Okay, thanks for explaining that those are legitimate wildcard characters, I will keep them as is then.

.. Retrieve the OIDC provider by running the following command:
+
[source,terminal]
----
$ oc get --raw=/.well-known/openid-configuration | jq -r '.issuer'
----
+

> // TODO: Can this be rewritten to use jsonpath or something else instead of jq? (To avoid making users install jq)

Author: @xingxingxia Is there any way to get the issuer value here without using jq? I ask because we try to avoid using jq in our procedures when possible, to avoid making users have to install jq if they don't already have it. 

When I tried, I saw that we can't use --raw and -o can't be used together. But I wasn't sure if there was any other way. Thoughts?

Reviewer: Below are the full output:
```
$ oc get --raw=/.well-known/openid-configuration
{"issuer":"https://xxxx.s3.ap-northeast-2.amazonaws.com","jwks_uri":"https://10.0.60.253:6443/openid/v1/jwks","response_types_supported":["id_token"],"subject_types_supported":["public"],"id_token_signing_alg_values_supported":["RS256"]}
```
`"issuer":"https://xxxxxxxx"` is OIDC URL. Note the whole `xxxxxxxx` after the `https://` is the OIDC provider name (not the OIDC provider arn. ARN has prefixed `arn:aws:.....<aws_account_id>`).

Author: Okay thanks. I don't see another easy way to just get the issuer value cleanly, so I will keep jq, and just add it as a prereq, thanks!

----
Copy the `<oidc_provider_arn>` from the output to use in the next step. For example, `<name>-oidc.s3.<aws_region>.amazonaws.com`.

.. Use the `ccoctl` tool to process the credentials request by running the following command:
+
[source,terminal]
----
$ ccoctl aws create-iam-roles \
    --name test-role --region=<aws_region> \
    --credentials-requests-dir=credentialsrequest-dir-aws \
    --identity-provider-arn <oidc_provider_arn> --output-dir=credrequests-ccoctl-output
----
+
.Example output
[source,terminal]
----
2023/05/15 18:10:34 Role arn:aws:iam::<aws_account_id>:role/test-role-test-namespace-aws-creds created
2023/05/15 18:10:34 Saved credentials configuration to: credrequests-ccoctl-output/manifests/cert-manager-aws-creds-credentials.yaml
2023/05/15 18:10:35 Updated Role policy for Role test-role-test-namespace-aws-creds
----
+
Copy the `<aws_role_arn>` from the output to use in the next step. For example, `arn:aws:iam::<aws_account_id>:role/test-role-test-namespace-aws-creds`.

.. Bind the service account with the role ARN by running the following command:
+
[source,terminal]
----
$ oc annotate sa/aws-provider eks.amazonaws.com/role-arn="<role_arn>"
----
+

> // TODO: Do we need to pass in a particular ns?

Author: @xingxingxia I was comparing this with the cert-manager procedure that followed similar steps, and they had passed in a namespace with their command. Do we need to specify a namespace here too?

Reviewer: The test case uses `oc new-project xxia-proj`, this command changes the context to the specific namespace, so all commands after this command are in the context of the specific namespace. So the test case does not explicitly set -n. You can set it:
```
oc annotate -n <test-namespace or my-namespace, unify it> sa/aws-provider eks.amazonaws.com/role-arn="<arn_role_arn> or <role_arn>, unify it"
```

Author: Thanks. So all of this should be done in the namespace then. I'll update to use either test-namespace or my-namespace throughout then

+
[source,terminal]
----
$ oc adm policy add-scc-to-user privileged -z csi-secrets-store-provider-aws -n openshift-cluster-csi-drivers
----

.. Create the provider resources by running the following command:
+
[source,terminal]
----
$ oc apply -f aws-provider.yaml
----

. Grant permission to allow the service account to read the AWS secret object:

.. Create a directory to contain the credentials request by running the following command:
+
[source,terminal]
----
$ mkdir credentialsrequest-dir-aws
----

.. Create a YAML file with the following configuration for the credentials request:
+
.Example `credentialsrequest.yaml` file
[source,yaml]
----
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: aws-provider-test
  namespace: openshift-cloud-credential-operator
spec:
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AWSProviderSpec
    statementEntries:
    - action:
      - "secretsmanager:GetSecretValue"
      - "secretsmanager:DescribeSecret"
      effect: Allow
      resource: "arn:*:secretsmanager:*:*:secret:testSecret-??????"
  secretRef:
    name: aws-creds

>     namespace: test-namespace

Reviewer: The namespace is the namespace that hosts the serviceproviderclass, the pods and the serviceaccount `aws-provider`. You use `test-namespace` in some places but use `my-namespace` in other places. This needs be changed.

Author: I went with "my-namespace". I also updated the naming of "test-role" to "my-role", just to keep a consistent naming scheme

  serviceAccountNames:
  - aws-provider
----
+
// TODO `resource` value: should we say something about this? I changed it from "secret:xxiaSecret" to "secret:testSecret". What are the "??????" supposed to be?

.. Retrieve the OIDC provider by running the following command:
+
[source,terminal]
----
$ oc get --raw=/.well-known/openid-configuration | jq -r '.issuer'
----
+
// TODO: Can this be rewritten to use jsonpath or something else instead of jq? (To avoid making users install jq)
+
.Example output
[source,terminal]
----

> https://<name>-oidc.s3.<aws_region>.amazonaws.com
> ----
> Copy the `<oidc_provider_arn>` from the output to use in the next step. For example, `<name>-oidc.s3.<aws_region>.amazonaws.com`.

Reviewer: As previous comment, in `"issuer":"https://xxxxxxxx"`, the whole `xxxxxxxx` after the `https://` is the OIDC provider name, so it does not make sense to write `<name>` as you type.
And the output does not contain the OIDC provider `arn`, it only contains the OIDC provider name.
And the OIDC provider name does not necessarily to be the format of `xxx-s3.xxx.amazonaws.com`. In ROSA STS cluster, the gotten OIDC name will appear like "d3gtsnipped.cloudfront.net/22nvlqvbbib5l8lj9snipped" (including a slash).
```suggestion
{"issuer":"https://xxxxxxxx",...}
----
Copy the OIDC provider name `xxxxxxxx` from the output to use in the next step.
```

Author: Okay I went with this:

```
.Example output
[source,terminal]
----
https://<oidc_provider_name>
----
Copy the OIDC provider name `<oidc_provider_name>` from the output to use in the next step.
```

And then updated the below command to use `<oidc_provider_name>`. Does that work?

```
$ ccoctl aws create-iam-roles \
    --name my-role --region=<aws_region> \
    --credentials-requests-dir=credentialsrequest-dir-aws \
    --identity-provider-arn <oidc_provider_name> --output-dir=credrequests-ccoctl-output
```

.. Use the `ccoctl` tool to process the credentials request by running the following command:
+
[source,terminal]
----
$ ccoctl aws create-iam-roles \
    --name test-role --region=<aws_region> \
    --credentials-requests-dir=credentialsrequest-dir-aws \
    --identity-provider-arn <oidc_provider_arn> --output-dir=credrequests-ccoctl-output
----
+
.Example output
[source,terminal]
----
2023/05/15 18:10:34 Role arn:aws:iam::<aws_account_id>:role/test-role-test-namespace-aws-creds created

> 2023/05/15 18:10:34 Saved credentials configuration to: credrequests-ccoctl-output/manifests/cert-manager-aws-creds-credentials.yaml

Reviewer: No `cert-manager`. You can hide line 218 with just ellipsis `...` given line 218 does not contain info used later.
```suggestion
...
```

Author: Good catch, copy paste from cert-manager file and didn't replace this one properly

2023/05/15 18:10:35 Updated Role policy for Role test-role-test-namespace-aws-creds
----
+
Copy the `<aws_role_arn>` from the output to use in the next step. For example, `arn:aws:iam::<aws_account_id>:role/test-role-test-namespace-aws-creds`.

.. Bind the service account with the role ARN by running the following command:
+
[source,terminal]
----

> $ oc annotate sa/aws-provider eks.amazonaws.com/role-arn="<role_arn>"

Reviewer: You use `<aws_role_arn>` in one line but use `<role_arn>` in the other line. Need be same.

          - "10000"
        volumeMounts:
        - name: secrets-store-inline
          mountPath: "/mnt/secrets-store"
          readOnly: true
      volumes:
        - name: secrets-store-inline
          csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: "my-aws-provider" <3>
----
<1> Specify the name for the deployment.
<2> Specify the namespace for the deployment. This must be the same namespace as the secret provider class.
<3> Specify the name of the secret provider class.

.. Create the `Deployment` object by running the following command:
+
[source,terminal]
----
$ oc create -f deployment.yaml
----

.Verification

* Verify that you can access the secrets from {secrets-store-provider} in the pod volume mount:

.. List the secrets in the pod mount:
+
[source,terminal]
----
$ oc exec busybox-<hash> -n my-namespace -- ls /mnt/secrets-store/
----
+
.Example output
[source,terminal]
----
testSecret
----

.. View a secret in the pod mount:
+
[source,terminal]
----
$ oc exec busybox-<hash> -n my-namespace -- cat /mnt/secrets-store/testSecret
----
+
.Example output
[source,terminal]
----

> {"username":"my-user", "password":"my-password"}

Reviewer: The secret value stems from the case's command `aws --region "$REGION" secretsmanager  create-secret --name xxiaSecret --secret-string '{"username":"memeuser", "password":"hunter2"}' --description "xxia test purpose"`. It can be any secret value. Will users be confuse where `{"username":"my-user", "password":"my-password"}` stems from? Is it better to be general like below?
```suggestion
<secret value>
```

Author: It is supposed to just be an example output, but I can definitely put just a replaceable value there if that would be less confusing
